app/models/service_provider.rb

```diff
@@ -1,17 +1,33 @@
 class ServiceProvider < ActiveRecord::Base
+  ISSUER_FORMAT_REGEXP = /\Aurn:gov:gsa:(SAML:2\.0|openidconnect)\.profiles:sp:sso:(.+):(.+)\z/
+
+  attr_writer :issuer_department, :issuer_app
+
   belongs_to :user
   belongs_to :agency
   belongs_to :user_group
 
   enum block_encryption: { 'aes256-cbc' => 1 }
   enum identity_protocol: { openid_connect: 0, saml: 1 }
 
-  validates :issuer, presence: true, uniqueness: true
+  # i18n-tasks-use t('activerecord.errors.models.service_provider.attributes.issuer.invalid')
+  validates :issuer, presence: true, uniqueness: true, format: { with: ISSUER_FORMAT_REGEXP }
+  validates :issuer_department, presence: true
+  validates :issuer_app, presence: true
   validates :agency, presence: true
 
   before_validation(on: %i(create update)) do
     self.attribute_bundle = attribute_bundle.reject(&:blank?) if attribute_bundle.present?
   end
+  before_validation :build_issuer
+
+  def issuer_department
```

Comment:
I'm not the biggest fan of bolting this onto the model, but adding a Form object to handle this started to feel heavy handed. I did the best I could to keep most of the logic in service objects.